File: programs/ch08/java/pos_tagging/src/util/ConfusionMatrix.java
package util;

import format.Word;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * @author pierre
 */
public class ConfusionMatrix {

    List<List<Word>> sentenceList;
    Map<String, Map<String, Integer>> confMat;

    public ConfusionMatrix(List<List<Word>> sentenceList) {
        this.sentenceList = sentenceList;
        confMat = new TreeMap<>();
    }

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        // TODO code application logic here
    }

    public float computeAccuracy() {
        int correct = 0, error = 0;
        for (List<Word> sentence : sentenceList) {
            for (int i = 1; i < sentence.size() - 1; i++) { // Skipping the ROOT index
                Word word = sentence.get(i);
                if (word.getPos().equals(word.getPpos())) {
                    correct++;
                } else {
                    error++;
                }
            }
        }
        return (float) correct / (float) (correct + error);
    }

    public Map<String, Map<String, Integer>> computeMatrix() {
        for (List<Word> sentence : sentenceList) {
            for (int i = 1; i < sentence.size() - 1; i++) { // Skipping the ROOT index
                Word word = sentence.get(i);
                if (confMat.get(word.getPos()) != null) {
                    if (confMat.get(word.getPos()).get(word.getPpos()) != null) {
                        int count = confMat.get(word.getPos()).get(word.getPpos());
                        count++;
                        confMat.get(word.getPos()).put(word.getPpos(), count);
                    } else {
                        confMat.get(word.getPos()).put(word.getPpos(), 1);
                    }
                } else {
                    Map<String, Integer> elem = new TreeMap<>();
                    elem.put(word.getPpos(), 1);
                    confMat.put(word.getPos(), elem);
                }
            }
        }
        System.out.println(confMat);
        return confMat;
    }

    public void print() {
        System.out.print("\t");
        for (String pos : confMat.keySet()) {
            System.out.print(pos + "\t");
        }
        System.out.println();
        for (String pos : confMat.keySet()) {
            System.out.print(pos + "\t");
            for (String ppos : confMat.keySet()) {
                if (confMat.get(pos).get(ppos) == null) {
                    System.out.print("0\t");
                } else {
                    System.out.print(confMat.get(pos).get(ppos) + "\t");
                }
            }
            System.out.println();
        }
    }

}

File: programs/ch08/python/hmm/confusion_matrix.py
__author__ = 'Pierre Nugues'


class ConfusionMatrix:

    def __init__(self, sentences, POS_key, relative_freqs=True):
        self.sentences = sentences
        self.conf_matrix = {}
        self.POS_key = POS_key
        self.relative_freqs = relative_freqs

    def compute_accuracy(self):
        correct = 0
        error = 0
        for sentence in self.sentences:
            for word in sentence:
                if word[self.POS_key] == word['PPOS']:
                    correct += 1
                else:
                    error += 1
        return correct / (correct + error)

    def compute_matrix(self):
        for sentence in self.sentences:
            for word in sentence:
                if word[self.POS_key] in self.conf_matrix:
                    if word['PPOS'] in self.conf_matrix[word[self.POS_key]]:
                        self.conf_matrix[word[self.POS_key]][word['PPOS']] += 1
                    else:
                        self.conf_matrix[word[self.POS_key]][word['PPOS']] = 1
                else:
                    self.conf_matrix[word[self.POS_key]] = {}
                    self.conf_matrix[word[self.POS_key]][word['PPOS']] = 1
        if self.relative_freqs:
            for key1 in self.conf_matrix:
                POS_count = sum(self.conf_matrix[key1][key2] for key2 in self.conf_matrix[key1])
                if POS_count == 0:
                    continue
                for key2 in self.conf_matrix[key1]:
                    self.conf_matrix[key1][key2] /= POS_count

    def print(self):
        print('\t')
        for pos in sorted(self.conf_matrix):
            print(pos, end='\t')
        print('\n')
        for pos in sorted(self.conf_matrix):
            print(pos, end='\t')
            for ppos in sorted(self.conf_matrix):
                if self.conf_matrix[pos].get(ppos) is None:
                    if self.relative_freqs:
                        print('{:.2f}'.format(0.0), end='\t')
                    else:
                        print('0', end='\t')
                else:
                    if self.relative_freqs:
                        print('{:.2f}'.format(self.conf_matrix[pos][ppos] * 100), end='\t')
                    else:
                        print(self.conf_matrix[pos][ppos], end='\t')
            print('\n')
